### ISSUE TITLE
Support for reading gameplay?

### ISSUE THREAD

**Issue by boombatower** [goto-bus-stop/recanalyst#1]
I scanned through documentation and some code, but I don't see anything to read the actual game play information instead of just stats and initial points.

A couple of questions if you feel like answering, otherwise I just need to do more research.
- Is the record game file writen in useful chunks that one could parse gameplay out of it as the game proceeds? (assuming written at some regular interval)
- Alternatively do you know of documentation (or better yet code) for acting as an observer client? I assume voobly has some sort of code or perhaps purely a packet proxy for its observer mode.

Basically, I am looking to read the actual state changes of the game (preferably during gameplay, but postprocessing would be a good first step) to supplement the UI.

Since you seem to have experience digging into some of this stuff just looking for some thoughts if you have time. Thanks.

**Comment by mtzander** [comment]
There isn't enough information in the rec to reconstruct anything terribly useful w.r.t. buildings and units. You won't be able to derive game state without accurately simulating the game.

That said, there is some useful data, and interesting metadata can be derived from it. I'm working on a spectator tool right now.

Getting spec data: Voobly spec mode writes voobly-spec.mgz as the game proceeds. You can read that file concurrently with the game. Packet traffic is compressed or encrypted, but contains extra data not in the rec, like enemy team chat.

**Comment by bowswung** [comment]
Just in case anyone looks at this in the future, I must respectfully disagree with @mtzander . It is possible to reconstruct a lot of information about buildings and units using solid inferences and without resorting to guesswork, as you can see from [this](https://github.com/bowswung/voobly-scraper/blob/d49d7a20d8555c03aa7f5d771c7b0eb8d9f1e4da/test/simHistory).

**Comment by danielsuter** [comment]
@bowswung They seem to have additional data from their client. From the specifications available there's no way to detect a death without having the engine simulation.

**Comment by IamFlea** [comment]
@danielsuter Hello, we have already solved that with bowswung. We made a tool called MAGAGA reader that logs data into flatbuffers which can be used in any language. You can ask him for getting this tool. 

Aside of detecting deaths, we also hooked damage dealing function. And we also speed the game simulation up to less than 5 minutes.

**Comment by danielsuter** [comment]
@IamFlea Hi, thank you for your answer. Do you run the actual Replay using age of Empires? Or do you have do use a mod when playing AOE2 which logs additional information?

**Comment by goto-bus-stop** [comment]
note that the file that bowswung linked is inferred _only_ from the recorded game. the tool Flea mentioned runs the actual replay using age of empires at high speed, with some code injected to log additional information.

**Comment by danielsuter** [comment]
@goto-bus-stop What confuses me that the log referred to in that comment is called **simHistory**. So that would mean deaths and everything can be inferred from the actual replay/recorded game in Aoe2De without actually replaying the game? The documentations I found only identified actions commands, so you would require the actual engine with the correct page to get data such as deaths or produced units.

**Comment by goto-bus-stop** [comment]
The recorded game file only records actions, but bowswung found ways to use that limited data to infer lots of things about the game. It's not 100% accurate, because it can't be. Most of the conversation about that is here: https://github.com/goto-bus-stop/recanalyst/issues/17
The code that does it is in the linked repo. The entry point is https://github.com/bowswung/voobly-scraper/blob/d49d7a20d8555c03aa7f5d771c7b0eb8d9f1e4da/src/Data/Mgz/Simulate.hs.

It's true that you need to actually run the engine for accurate data, which AFAIK is currently only possible in AoC with UserPatch using the MAGAGA tool.
